File: app/src/main/java/com/castelijns/mmdemo/photos/PhotoSection.java
package com.castelijns.mmdemo.photos;

import android.content.Context;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.castelijns.mmdemo.R;
import com.castelijns.mmdemo.models.Photo;

import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionParameters;
import io.github.luizgrp.sectionedrecyclerviewadapter.StatelessSection;

public class PhotoSection extends StatelessSection {

    private ItemClickListener itemClickListener;

    private Context context;
    private int albumId;
    private List<Photo> photos;
    private int itemWidth;

    PhotoSection(Context context, int albumId, List<Photo> photos, int itemWidth) {
        super(new SectionParameters.Builder(R.layout.photos_section_item)
                .headerResourceId(R.layout.photos_section_header)
                .build());

        this.context = context;
        this.albumId = albumId;
        this.photos = photos;
        this.itemWidth = itemWidth;
    }

    @Override
    public int getContentItemsTotal() {
        return photos.size();
    }

    @Override
    public RecyclerView.ViewHolder getItemViewHolder(View view) {
        return new ItemViewHolder(view);
    }

    @Override
    public void onBindItemViewHolder(RecyclerView.ViewHolder holder, int position) {
        final Photo photo = photos.get(position);
        final ItemViewHolder itemHolder = (ItemViewHolder) holder;

        String thumbnailurl = photo.getThumbnailUrl();

        itemHolder.ivPhoto.getLayoutParams().width = itemWidth;
        itemHolder.ivPhoto.getLayoutParams().height= itemWidth;
        itemHolder.ivPhoto.requestLayout();

        Glide.with(context)
                .load(thumbnailurl)
                .into(itemHolder.ivPhoto);

        itemHolder.itemView.setOnClickListener(v -> {
            if (itemClickListener != null) {
                itemClickListener.onClick(photo, itemHolder.ivPhoto);
            }
        });

        // For shared element transitions.
        ViewCompat.setTransitionName(itemHolder.ivPhoto, String.format("%s.%s",
                "title", photo.getTitle()));
    }

    @Override
    public RecyclerView.ViewHolder getHeaderViewHolder(View view) {
        return new HeaderViewHolder(view);
    }

    @Override
    public void onBindHeaderViewHolder(RecyclerView.ViewHolder holder) {
        HeaderViewHolder headerHolder = (HeaderViewHolder) holder;

        headerHolder.tvTitle.setText(String.format(Locale.getDefault(),
                "%s %d", context.getString(R.string.album), albumId));
    }

    class HeaderViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.tv_title)
        TextView tvTitle;

        HeaderViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }
    }

    class ItemViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.iv_photo)
        ImageView ivPhoto;

        ItemViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }
    }

    void setItemClickListener(ItemClickListener listener) {
        itemClickListener = listener;
    }

    interface ItemClickListener {
        void onClick(Photo photo, ImageView ivPhoto);
    }
}

File: app/src/main/kotlin/com/castelijns/mmdemo/photos/PhotoSection.kt
package com.castelijns.mmdemo.photos

import android.content.Context
import android.view.View
import android.widget.ImageView
import androidx.core.view.ViewCompat
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.castelijns.mmdemo.R
import com.castelijns.mmdemo.models.Photo
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionParameters
import io.github.luizgrp.sectionedrecyclerviewadapter.StatelessSection
import kotlinx.android.synthetic.main.photos_section_header.view.*
import kotlinx.android.synthetic.main.photos_section_item.view.*
import java.util.*

class PhotoSection(
        private val context: Context,
        private val albumId: Int,
        private val albumName: String,
        private val photos: List<Photo>,
        private val itemWidth: Int
) : StatelessSection(SectionParameters.Builder(R.layout.photos_section_item)
        .headerResourceId(R.layout.photos_section_header)
        .build()) {

    private var itemClickListener: ItemClickListener? = null

    override fun getContentItemsTotal(): Int {
        return photos.size
    }

    override fun getItemViewHolder(view: View): RecyclerView.ViewHolder {
        return ItemViewHolder(view)
    }

    override fun onBindItemViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val photo = photos[position]

        val thumbnailUrl = photo.thumbnailUrl

        holder.itemView.iv_photo.apply {
            layoutParams.width = itemWidth
            layoutParams.height = itemWidth
            requestLayout()
        }

        Glide.with(context)
                .load(thumbnailUrl)
                .into(holder.itemView.iv_photo)

        holder.itemView.setOnClickListener {
            if (itemClickListener != null) {
                itemClickListener!!.onClick(photo, holder.itemView.iv_photo)
            }
        }

        // For shared element transitions.
        ViewCompat.setTransitionName(holder.itemView.iv_photo, String.format("%s.%s",
                "title", photo.title))
    }

    override fun getHeaderViewHolder(view: View): RecyclerView.ViewHolder {
        return HeaderViewHolder(view)
    }

    override fun onBindHeaderViewHolder(holder: RecyclerView.ViewHolder) {
        val headerHolder = holder as HeaderViewHolder

        headerHolder.itemView.tv_id.text = String.format(Locale.getDefault(),
                "%s %d", context.getString(R.string.album), albumId)
        headerHolder.itemView.tv_title.text = albumName
    }

    class HeaderViewHolder(view: View) : RecyclerView.ViewHolder(view)

    class ItemViewHolder(view: View) : RecyclerView.ViewHolder(view)

    fun setItemClickListener(listener: ItemClickListener) {
        itemClickListener = listener
    }

    interface ItemClickListener {
        fun onClick(photo: Photo, ivPhoto: ImageView)
    }
}
